lib/rules/no-reassign-constant.js

```diff
@@ -0,0 +1,141 @@
+        var isWrite = (ref.isWrite() || !ref.isReadOnly());
+
+        if (!isWrite && isDeleted(ref.identifier.parent)) {
+            isWrite = true;
+        }
+
+        return isWrite;
+    }
+
+    /**
+     * Report a reference as invalid
+     *
+     * @param {String} name - the variables name
+     * @param {Reference} ref - the reference to report
+     * @returns {void}
+     * @private
+    */
+    function report(name, ref) {
+        context.report(ref.identifier, MESSAGE, {name: name});
+    }
+
+    /**
+     * Report a reference as invalid
+     *
+     * @param {Scope} scope - an escope object
+     * @param {Variable} variable - variable to apply references too
+     * @returns {Variable} returns the variable with references
+     * @private
+    */
+    function transformGlobalVariables(scope, variable) {
+        scope.references.forEach(function(ref) {
+            if (ref.identifier.name === variable.name) {
+                variable.references.push(ref);
+            }
+        });
+
+        return variable;
+    }
+
+    /**
+     * Checks for and reports reassigned constants
+     *
+     * @param {Scope} scope - an escope Scope object
+     * @returns {void}
+     * @private
+    */
+    function checkScope(scope) {
+        var variables = scope.variables;
+
+        if (!scope.functionExpressionScope) {
+            for (var i = 0, l = variables.length; i < l; ++i) {
+                // skip function expression names
+                if (variables[i].eslintUsed) {
+                    continue;
+                }
+
+                // skip implicit "arguments" variable
+                if (scope.type === "function" && variables[i].name === "arguments" && variables[i].identifiers.length === 0) {
+                    continue;
+                }
+
+                var def = variables[i].defs[0];
+
+                // only constants
+                if (!def || def.kind !== "const") {
+                    continue;
+                }
+
+                var variable = variables[i];
+                var references = variable.references;
+                var name = variable.name;
+                var assignments = references.filter(isReassignment);
+
+                if (assignments.length > 1) {
+                    // remove original assigment
+                    assignments.shift();
+
+                    assignments.forEach(report.bind(null, name));
+                }
+            }
+        }
+
+        scope.childScopes.forEach(checkScope);
+    }
+
+    return {
+        "Program:exit": function() {
+            var scope = context.getScope();
+
+            if (scope.type === "global") {
+                scope = {
+                    childScopes: scope.childScopes,
+                    variables: scope.variables.map(transformGlobalVariables.bind(null, scope))
+                };
+            }
+
+            if (context.ecmaFeatures.blockBindings) {
+                checkScope(scope);
```

Comment:
Thanks for the feedback, first time using eslint and the info you provided has helped a lot. As for the approach I originally was using ExpressionStatement to get UnaryExpressions and UpdateExpressions but I assumed the recursion would better performance. 

Here is the reasoning, please let me know if I'm mistaken, I assumed that at each level in the scope we are likely to have an assignment, update or update and possibly multiple. Since the expression doesn't expose the definition we need to take the variable name and loop through the variables in the scope. By looping through the defined variables instead of when they are changed, we decrease the number of times we have to loop through a scopes variables. Since we limit looping through references only to const, it should in theory work better?